### PR TITLE
Add online installer script

### PR DESCRIPTION
**Pull request by ali-behjati**
This PR adds an online installer scripts which will enable one line installation of tps using:

```
bash -c "$(curl -fsSL https://raw.githubusercontent.com/ioi-2017/tps/master/online-installer/install.sh)"
```

It will also clone the tps repo to `~/.local/share/tps` (or configured `$TPS_LOCAL_REPO`) to be able to use its data (such as `scripts`) in the future.
 
I tested this script in Mac and Linux (via docker) in both bash and zsh.

### CODE REVIEW

File: online-installer/install.sh [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+# Make sure important variables exist if not already defined
+#
+# $USER is defined by login(1) which is not always executed (e.g. containers)
+# POSIX: https://pubs.opengroup.org/onlinepubs/009695299/utilities/id.html
+USER=${USER:-$(id -u -n)}
+# $HOME is defined at the time of login, but it could be unset. If it is unset,
+# a tilde by itself (~) will not be expanded to the current user's home directory.
+# POSIX: https://pubs.opengroup.org/onlinepubs/009696899/basedefs/xbd_chap08.html#tag_08_03
+HOME="${HOME:-$(getent passwd $USER 2>/dev/null | cut -d: -f6)}"
```
Comment (mirjalali) [inline online-installer/install.sh]:
```suggestion
HOME="${HOME:-$(getent passwd $USER 2> "/dev/null" | cut -d: -f6)}"
```

File: online-installer/install.sh [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+TPS="${TPS:-$HOME/.local/share/tps}"
+
+# Tps cli file path. It will be a symlink from the code repository and the directory should be added to the path.
+TPS_BIN="${TPS_BIN:-$HOME/.local/bin/tps}"
+REPO=${REPO:-ioi-2017/tps}
+REMOTE=${REMOTE:-https://github.com/${REPO}.git}
+BRANCH=${BRANCH:-master}
+
+command_exists() {
+	command -v "$@" >/dev/null 2>&1
```
Comment (mirjalali) [inline online-installer/install.sh]:
```suggestion
	command -v "$@" &> "/dev/null"
```

File: online-installer/install.sh [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+HOME="${HOME:-$(eval echo ~$USER)}"
+
+# Directory of TPS code repository
+TPS="${TPS:-$HOME/.local/share/tps}"
+
+# Tps cli file path. It will be a symlink from the code repository and the directory should be added to the path.
+TPS_BIN="${TPS_BIN:-$HOME/.local/bin/tps}"
+REPO=${REPO:-ioi-2017/tps}
+REMOTE=${REMOTE:-https://github.com/${REPO}.git}
+BRANCH=${BRANCH:-master}
```
Comment (mirjalali) [inline online-installer/install.sh]: I suggest inlining `REPO` into `REMOTE` as some users might want to use non-github repositories.

Additionally, the name of these environment variables seems to be too general and might be defined for other purposes which may cause problems here.
I recommend names such as `TPS_REMOTE_REPO` and `TPS_REMOTE_BRANCH`.

The name of variable `TPS` is also too general and ambiguous.
Something like `TPS_LOCAL_REPO` seems to be better.

Comment (ali-behjati) [reply online-installer/install.sh]: Makes sense in general.

About your first comment, one can set `REMOTE` entirely and it's easier to keep it as it is right now.

File: online-installer/install.sh [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+	if user_can_sudo; then
+		echo "You might need to enter your password for installation."
+		sudo -k bash "install-tps.sh"
+	else
+		bash "install-tps.sh"
+	fi 
+
+	# Checking if installation was successful
+	if [ $? -ne 0 ]; then
```
Comment (mirjalali) [inline online-installer/install.sh]:
```suggestion
		sudo -k bash "install-tps.sh"
		install_exit_code=$?
	else
		bash "install-tps.sh"
		install_exit_code=$?
	fi 

	# Checking if installation was successful
	if [ "${install_exit_code}" -ne "0" ]; then
```

File: online-installer/install.sh [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+	command_exists "git" || {
+		fmt_error "git is not installed"
+		exit 1
+	}
+
+	ostype=$(uname)
+	if [ -z "${ostype%CYGWIN*}" ] && git --version | grep -q msysgit; then
+		fmt_error "Windows/MSYS Git is not supported on Cygwin"
+		fmt_error "Make sure the Cygwin git package is installed and is first on the \$PATH"
```
Comment (mirjalali) [inline online-installer/install.sh]: This code makes `Error: ` be printed twice. Is this intentional?

File: online-installer/install.sh [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+	
+	cd "${TPS_LOCAL_REPO}"
+
+	install_exit_code=0
+	if user_can_sudo; then
+		echo "You might need to enter your password for installation."
+		sudo -k bash "install-tps.sh" || install_exit_code=$?
+	else
+		bash "install-tps.sh"
+		install_exit_code=$? || install_exit_code=$?
```
Comment (mirjalali) [inline online-installer/install.sh]: Bug in the latest change

File: online-installer/install.sh (hunk 5 above)
Comment (ali-behjati) [reply online-installer/install.sh]: Thanks. Fixed.

File: online-installer/install.sh (hunk 6 above)
Comment (ali-behjati) [reply online-installer/install.sh]: Thanks. Fixed.

File: online-installer/install.sh [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+	command_exists "git" || {
+		fmt_error "git is not installed"
+		exit 1
+	}
+
+	ostype=$(uname)
+	if [ -z "${ostype%CYGWIN*}" ] && git --version | grep -q msysgit; then
+		fmt_error "Windows/MSYS Git is not supported on Cygwin." \
+			"Make sure the Cygwin git package is installed and is first on the \$PATH"
```
Comment (mirjalali) [inline online-installer/install.sh:90]:
```suggestion
		fmt_error "\
Windows/MSYS Git is not supported on Cygwin.
Make sure the Cygwin git package is installed and is first on the \$PATH."
```

Comment (ali-behjati) [reply online-installer/install.sh:90]: I think it's better for the error to be in one line.

**Review by mirjalali — APPROVED** [review]

**Pull request merged**